From f4da5523eac3c0bb88efdf14002c5d26918de910 Mon Sep 17 00:00:00 2001
From: "sameer.mulla" <sameer.mulla@lge.com>
Date: Wed, 30 Jan 2019 15:17:11 +0530
Subject: [PATCH] Added dbus signal for MediaPlayRequest

:Release Notes:
Added dbus signal for MediaPlayRequest and MediaMetaRequest

:Detailed Notes:
SIL needs information about MediaPlayRequest and MediaMetaRequest
so added above signals to inform about request to SIL

:Testing Performed:
Builded and Tested

:QA Notes:

:Issues Addressed:
[PLAT-73147] Implement avrcp/awaitMediaMetaDataRequest

Change-Id: Ib225395809a3b53f1c494ac4e21beabc47f7f906
---
Upstream-Status: Inappropriate [webos specific]

 profiles/audio/avrcp.c |  2 ++
 profiles/audio/media.c |  2 +-
 src/device.c           | 24 +++++++++++++++++++++++-
 src/device.h           |  3 +++
 4 files changed, 29 insertions(+), 2 deletions(-)

diff --git a/profiles/audio/avrcp.c b/profiles/audio/avrcp.c
index b5e6548e6..b6a7e4cdc 100644
--- a/profiles/audio/avrcp.c
+++ b/profiles/audio/avrcp.c
@@ -1259,6 +1259,7 @@ static uint8_t avrcp_handle_get_element_attributes(struct avrcp *session,
 		 */
 		attr_ids = player_list_metadata(player);
 		len = g_list_length(attr_ids);
+		media_metadata_request(session->dev);
 	} else {
 		unsigned int i;
 		for (i = 0, len = 0, attr_ids = NULL; i < nattr; i++) {
@@ -1522,6 +1523,7 @@ static uint8_t avrcp_handle_get_play_status(struct avrcp *session,
 
 	pdu->params_len = cpu_to_be16(9);
 
+	media_play_request(session->dev);
 	return AVC_CTYPE_STABLE;
 }
 
diff --git a/profiles/audio/media.c b/profiles/audio/media.c
index 746e538fc..c434043e1 100644
--- a/profiles/audio/media.c
+++ b/profiles/audio/media.c
@@ -2024,7 +2024,7 @@ static uint32_t media_player_get_position(void *user_data)
 	sec = (uint32_t) timedelta;
 	msec = (uint32_t) ((timedelta - sec) * 1000);
 
-	return mp->position + sec * 1000 + msec;
+	return mp->position;
 }
 
 static uint32_t media_player_get_duration(void *user_data)
diff --git a/src/device.c b/src/device.c
index 83d5b8aaf..6147105e6 100644
--- a/src/device.c
+++ b/src/device.c
@@ -2347,6 +2347,22 @@ void recieve_passThrough_commands(struct btd_device *dev, const char* code, cons
 					DEVICE_INTERFACE, "KeyCode");
 }
 
+void media_play_request(struct btd_device *dev)
+{
+	DBG("");
+	g_dbus_emit_signal(dbus_conn, dev->path,
+				DEVICE_INTERFACE, "MediaPlayRequest",
+				DBUS_TYPE_INVALID);
+}
+
+void media_metadata_request(struct btd_device *dev)
+{
+	DBG("");
+	g_dbus_emit_signal(dbus_conn, dev->path,
+			DEVICE_INTERFACE, "MediaMetaRequest",
+			DBUS_TYPE_INVALID);
+}
+
 static void add_service_data(void *data, void *user_data)
 {
 	struct eir_sd *sd = data;
@@ -3439,6 +3455,12 @@ static DBusMessage *get_service_records(DBusConnection *conn, DBusMessage *msg,
 	return reply;
 }
 
+static const GDBusSignalTable device_avrcp_media_request_signal[] = {
+	{ GDBUS_SIGNAL("MediaPlayRequest", NULL) },
+	{ GDBUS_SIGNAL("MediaMetaRequest", NULL) },
+	{ }
+};
+
 static const GDBusMethodTable device_methods[] = {
 	{ GDBUS_ASYNC_METHOD("Disconnect", NULL, NULL, dev_disconnect) },
 	{ GDBUS_ASYNC_METHOD("Connect", NULL, NULL, dev_connect) },
@@ -4536,7 +4558,7 @@ static struct btd_device *device_new(struct btd_adapter *adapter,
 
 	if (g_dbus_register_interface(dbus_conn,
 					device->path, DEVICE_INTERFACE,
-					device_methods, NULL,
+					device_methods, device_avrcp_media_request_signal,
 					device_properties, device,
 					device_free) == FALSE) {
 		error("Unable to register device interface for %s", address);
diff --git a/src/device.h b/src/device.h
index 5bec61ba1..51b9f801b 100644
--- a/src/device.h
+++ b/src/device.h
@@ -216,3 +216,6 @@ void btd_device_foreach_service_data(struct btd_device *dev,
 					void *data);
 
 void recieve_passThrough_commands(struct btd_device *dev, const char* code, const char* state);
+
+void media_play_request(struct btd_device *dev);
+void media_metadata_request(struct btd_device *dev);
